core/services/vrf/listener_v2.go

```diff
@@ -217,18 +218,22 @@ func (lsn *listenerV2) processPendingVRFRequests() {
 			return
 		}
 		startBalance := sub.Balance
-		lsn.processRequestsPerSub(fromAddress.Address(), startBalance, maxGasPrice, reqs)
+		lsn.processRequestsPerSub(subID, fromAddress.Address(), startBalance, maxGasPrice, reqs)
 	}
 	lsn.pruneConfirmedRequestCounts()
 }
 
-func MaybeSubtractReservedLink(l logger.Logger, db *gorm.DB, fromAddress common.Address, startBalance *big.Int) (*big.Int, error) {
+// MaybeSubtractReservedLink figures out how much LINK is reserved for other VRF requests that
+// have not been fully confirmed yet on-chain, and subtracts that from the given startBalance,
+// and returns that value if there are no errors.
+func MaybeSubtractReservedLink(l logger.Logger, db *gorm.DB, fromAddress common.Address, startBalance *big.Int, subID uint64) (*big.Int, error) {
 	var reservedLink string
-	err := db.Raw(`SELECT SUM(CAST(meta->>'MaxLink' AS NUMERIC(78, 0))) 
-					FROM eth_txes
-					WHERE meta->>'MaxLink' IS NOT NULL
-					AND (state <> 'fatal_error' AND state <> 'confirmed' AND state <> 'confirmed_missing_receipt') 
-					GROUP BY from_address = ?`, fromAddress).Scan(&reservedLink).Error
+	err := db.Raw(`SELECT SUM(CAST(meta->>'MaxLink' AS NUMERIC(78, 0)))
+				   FROM eth_txes
+				   WHERE meta->>'MaxLink' IS NOT NULL
+				   AND CAST(meta->>'SubId' AS NUMERIC) = ?
+				   AND (state <> 'fatal_error' AND state <> 'confirmed' AND state <> 'confirmed_missing_receipt')
+				   GROUP BY from_address = ?`, subID, fromAddress).Scan(&reservedLink).Error
```

Comment:
Have you run EXPLAIN ANALYZE with this query on a real world data set? What were the results?

Reply:
Not yet, will be done in another PR and ticket: https://app.shortcut.com/chainlinklabs/story/21433/benchmark-and-optimize-maybesubtractreservedlink?cf_workflow=500000005&ct_workflow=all.

Reply:
Actually, I've ended up improving it in this PR. Here are my findings.

## Context
* `eth_txes` table has slightly more than 1 million rows, all of which have a from_address equal to the same value (this is likely how it'll be in prod, node ops' addresses will occupy the from_address column)
* All of these transactions are `unconfirmed` (so that the query where clause kicks in)
* Ran a `vacuum analyze` prior to running these queries to optimize space usage by the table.

## Query 1
Query:

```sql
explain analyze SELECT SUM(CAST(meta->>'MaxLink' AS NUMERIC(78, 0)))
FROM eth_txes
WHERE meta->>'MaxLink' IS NOT NULL
	AND CAST(meta->>'SubId' AS NUMERIC) = 1
	AND (state <> 'fatal_error' AND state <> 'confirmed' AND state <> 'confirmed_missing_receipt')
GROUP BY from_address = '\xb5de2fb686baa26047f047aa0ec766d5cb97b0d1';
```
The values given for SubId and from_address are just examples, just from the data I generated and inserted into my local DB instance.

Here's the query plan and output of `explain analyze` for the above query:

```
                                                                                                                                        QUERY PLAN                                                                                                                                        
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Finalize GroupAggregate  (cost=122949.77..122987.17 rows=2 width=33) (actual time=2659.964..2660.069 rows=1 loops=1)
   Group Key: ((from_address = '\xb5de2fb686baa26047f047aa0ec766d5cb97b0d1'::bytea))
   ->  Gather Merge  (cost=122949.77..122987.11 rows=4 width=33) (actual time=2659.958..2660.064 rows=3 loops=1)
         Workers Planned: 2
         Workers Launched: 2
         ->  Partial GroupAggregate  (cost=121949.74..121986.63 rows=2 width=33) (actual time=2396.208..2396.209 rows=1 loops=3)
               Group Key: ((from_address = '\xb5de2fb686baa26047f047aa0ec766d5cb97b0d1'::bytea))
               ->  Sort  (cost=121949.74..121955.01 rows=2106 width=240) (actual time=2110.108..2385.434 rows=30512 loops=3)
                     Sort Key: ((from_address = '\xb5de2fb686baa26047f047aa0ec766d5cb97b0d1'::bytea))
                     Sort Method: external merge  Disk: 8040kB
                     Worker 0:  Sort Method: external merge  Disk: 7400kB
                     Worker 1:  Sort Method: external merge  Disk: 7400kB
                     ->  Parallel Seq Scan on eth_txes  (cost=0.00..121833.49 rows=2106 width=240) (actual time=101.551..362.951 rows=30512 loops=3)
                           Filter: (((meta ->> 'MaxLink'::text) IS NOT NULL) AND (state <> 'fatal_error'::eth_txes_state) AND (state <> 'confirmed'::eth_txes_state) AND (state <> 'confirmed_missing_receipt'::eth_txes_state) AND (((meta ->> 'SubId'::text))::numeric = '1'::numeric))
                           Rows Removed by Filter: 306154
 Planning Time: 23.895 ms
 Execution Time: 2666.949 ms
(17 rows)
```

Total execution time around 2.6 seconds.

## Query 2

This is a query that should be equivalent to the prior query, just moving the `from_address` check to the `where` clause instead:

```sql
explain analyze select sum(cast(meta->>'MaxLink' as numeric(78, 0)))
from eth_txes
where meta->>'MaxLink' is not null
	and cast(meta->>'SubId' as numeric) = 1
	and (state <> 'fatal_error' AND state <> 'confirmed' AND state <> 'confirmed_missing_receipt')
	and from_address = '\xb5de2fb686baa26047f047aa0ec766d5cb97b0d1'
group by from_address ;
```

I believe this is also more correct, since Query 1 is susceptible to a subtle bug where it will return two rows in the result set if there is more than one from_address in the `eth_txes` table (unsure of why this is the case).

Here's the query plan:

```
                                                                                                                                                                         QUERY PLAN                                                                                                                                                                          
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Finalize GroupAggregate  (cost=1000.00..123918.24 rows=1 width=53) (actual time=564.795..568.059 rows=1 loops=1)
   Group Key: from_address
   ->  Gather  (cost=1000.00..123918.21 rows=2 width=53) (actual time=564.635..568.052 rows=3 loops=1)
         Workers Planned: 2
         Workers Launched: 2
         ->  Partial GroupAggregate  (cost=0.00..122918.01 rows=1 width=53) (actual time=547.601..547.602 rows=1 loops=3)
               Group Key: from_address
               ->  Parallel Seq Scan on eth_txes  (cost=0.00..122886.41 rows=2106 width=260) (actual time=134.631..540.264 rows=30512 loops=3)
                     Filter: (((meta ->> 'MaxLink'::text) IS NOT NULL) AND (state <> 'fatal_error'::eth_txes_state) AND (state <> 'confirmed'::eth_txes_state) AND (state <> 'confirmed_missing_receipt'::eth_txes_state) AND (from_address = '\xb5de2fb686baa26047f047aa0ec766d5cb97b0d1'::bytea) AND (((meta ->> 'SubId'::text))::numeric = '1'::numeric))
                     Rows Removed by Filter: 306154
 Planning Time: 34.971 ms
 Execution Time: 568.237 ms
(12 rows)
```

Total query execution time hovered around 0.55 seconds, which is ~5x (4.6x) faster than Query 1.

## Query 2 with index on `meta->>'SubId'`

If we add the following index:

```sql
create index on eth_txes(CAST(meta->>'SubId' AS numeric))
```

and execute Query 2 above, we get the following query plan:

```
                                                                                                                                       QUERY PLAN                                                                                                                                       
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 GroupAggregate  (cost=95.56..16929.78 rows=1 width=53) (actual time=234.008..234.010 rows=1 loops=1)
   Group Key: from_address
   ->  Bitmap Heap Scan on eth_txes  (cost=95.56..16854.40 rows=5025 width=260) (actual time=25.374..214.547 rows=91537 loops=1)
         Recheck Cond: (((meta ->> 'SubId'::text))::numeric = '1'::numeric)
         Filter: (((meta ->> 'MaxLink'::text) IS NOT NULL) AND (state <> 'fatal_error'::eth_txes_state) AND (state <> 'confirmed'::eth_txes_state) AND (state <> 'confirmed_missing_receipt'::eth_txes_state) AND (from_address = '\xb5de2fb686baa26047f047aa0ec766d5cb97b0d1'::bytea))
         Heap Blocks: exact=41590
         ->  Bitmap Index Scan on eth_txes_numeric_idx  (cost=0.00..94.30 rows=5050 width=0) (actual time=20.914..20.914 rows=91537 loops=1)
               Index Cond: (((meta ->> 'SubId'::text))::numeric = '1'::numeric)
 Planning Time: 0.927 ms
 Execution Time: 234.114 ms
(10 rows)
```

Execution times ranges between 0.2s and 0.4s. In the best case scenario, we see a 13.3x speedup on Query 1, though only a 2.84x speedup on Query 2 without the index.

Reply:
Nice, let's add the index

Reply:
I've updated the query to group by subscriber ID. It ends up using the `idx_eth_txes_state_from_address_evm_chain_id` index and is quite fast. Here's the query plan:

```
                                                                                              QUERY PLAN                                                                                               
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 GroupAggregate  (cost=53614.50..53615.63 rows=11 width=64) (actual time=7.361..7.362 rows=1 loops=1)
   Group Key: ((meta ->> 'SubId'::text))
   ->  Sort  (cost=53614.50..53614.64 rows=55 width=271) (actual time=7.100..7.140 rows=993 loops=1)
         Sort Key: ((meta ->> 'SubId'::text))
         Sort Method: quicksort  Memory: 537kB
         ->  Index Scan using idx_eth_txes_state_from_address_evm_chain_id on eth_txes  (cost=0.41..53612.91 rows=55 width=271) (actual time=0.053..6.246 rows=993 loops=1)
               Index Cond: (evm_chain_id = '1337'::numeric)
               Filter: (((meta ->> 'MaxLink'::text) IS NOT NULL) AND (state = ANY ('{unconfirmed,unstarted,in_progress}'::eth_txes_state[])) AND (((meta ->> 'SubId'::text))::numeric = '1'::numeric))
               Rows Removed by Filter: 10563
 Planning Time: 21.049 ms
 Execution Time: 7.508 ms
```

On a DB with 1M total rows for an address and around 1% of those being unconfirmed/unstarted/in_progress.